### ISSUE TITLE
[BUG] Incorrect molecule IDs from fix reaxff/species

### ISSUE THREAD

**Issue by navlalli** [lammps/lammps#4373]
**Summary**

The molecule IDs returned by `fix reaxff/species` are sometimes incorrect.

**LAMMPS Version and Platform**

LAMMPS version: LAMMPS (29 Aug 2024 - Update 1)
Platform: Ubuntu 20.04.2 LTS (fossa-ditto X84)
This error also occurs on Red Hat Enterprise Linux 8.5 (Ootpa).

**Steps to Reproduce**

I have attached a minimal working example (`in.test`) with the data file and the reaxff force field file needed to reproduce this issue (see [test.zip](https://github.com/user-attachments/files/17680105/test.zip)). It takes 14s to run with `mpirun -np 16 ~/lammps-29Aug2024/build/lmp -in in.test -log log.run` on my system. 

The data file contains 30 hydrocarbon molecules, which each contain 38 atoms, and 555 oxygen molecules. We can see from `dump.run` that the molecule IDs (the fourth column) are from 1 to 585 for timestep 230, with all molecules being assigned the correct ID. For instance, the first 38 atoms are assigned mol ID 1 and the last two atoms (an oxygen molecule) are assigned mol ID 585.

In contrast, there are errors in the molecule IDs for timestep 1150. For instance, the dump file contains
``` 
atom ID, element, q, mol ID, x, y, z
1997 O -0.00372084 459 -7.62194 -9.64018 13.2621
1998 O -0.0155002 2217 -7.11975 -9.47061 14.3768
...
2249 O -0.0068809 585 0.0830279 -3.45521 9.10559
2250 O -0.0172618 459 -0.990076 -3.51245 9.65433
```
Evidently from the positions, the oxygen with atom ID 1997 is not bonded with the oxygen with atom ID 2250. It is also unclear why the oxygen with ID 1998 is assigned a molecule ID of 2217 when there are only 585 molecules (as also confirmed by the species file output by `fix reaxff/species`). I'd have expected the oxygens with atom IDs 1997 and 1998 to have been assigned mol ID 459, as in the other timesteps in the dump file. Similarly, I'd have expected the oxygens with atom IDs 2249 and 2250 to have been assigned mol ID 585.

I should note that this error does not occur with this example when running with only one core.

I'd be happy to provide more examples if needed. I have a lot of dump files in which the molecule IDs are randomly wrong, i.e., at some time steps they appear to be correct and in others they appear randomly messed up.

**Comment by mkanski** [comment]
It seems that the issue comes from your `neigh_modify` line. If I comment it out, I get the correct mol IDs. Can you confirm?

**Comment by navlalli** [comment]
This is true, but `neigh_modify every 5 delay 0 check yes` is not the cause of the problem. Attached is an example that reproduces this issue without using `neigh_modify` ([test2.zip](https://github.com/user-attachments/files/17777882/test2.zip)).

Looking at timestep 405, we see lines likes
```
atom ID, element, q, mol ID, x, y, z
889 H 0.0366273 1723 7.75494 6.94692 14.1431
...
1597 O -0.0188129 18 -7.28339 -13.0237 -0.0357332
1598 O 0.0114033 259 -6.47419 -13.8796 -0.253195
1599 O -0.0210487 260 0.852145 -22.0504 -26.3577
1600 O -0.0254196 1973 0.0236733 -22.9614 -26.4044
...
2219 O -0.0182742 30 20.4717 -1.54718 23.9615
...
2249 O 0.000763825 344 -0.272768 -3.8861 9.98009
2250 O -0.0357539 30 -1.19749 -3.43882 9.36559
```

**Comment by mkanski** [comment]
I cannot reproduce this with the current LAMMPS version from the `develop` branch. Can you post output of the `lmp -h` command?

Also, your output looks somehow incorrect - the molID for atom 1600 is equal to 1973 which is too high - the numbers should be consecutive from 1 to the number of molecules, so it would mean that almost every atom is a different molecule.
Can you also post last two lines from file `species.run`? Attaching the whole log file would also be helpful.

In my case it is:
```
#  Timestep    No_Moles    No_Specs      C12H26          O2
        500         585           2          30         555
```

**Comment by navlalli** [comment]
To confirm, you get the correct molecule IDs **at timestep 405** when running `in.test` in [test2.zip](https://github.com/user-attachments/files/17778894/test2.zip)? I should highlight that the molecule IDs appear to be correct for all other timesteps.

Everything, including the dump file is in [test-full.zip](https://github.com/user-attachments/files/17778957/test-full.zip). You can see that the output in `species.run` is the same for all timesteps.
```
#  Timestep    No_Moles    No_Specs      C12H26          O2
        All         585           2          30         555
```

Here is part of the output from `lmp -h`
```
Large-scale Atomic/Molecular Massively Parallel Simulator - 29 Aug 2024 - Update 1

Usage example: /home/nav/lammps-29Aug2024/build/lmp -var t 300 -echo screen -in in.alloy

List of command line options supported by this LAMMPS executable:

-echo none/screen/log/both  : echoing of input script (-e)
-help                       : print this help message (-h)
-in none/filename           : read input from file or stdin (default) (-i)
-kokkos on/off ...          : turn KOKKOS mode on or off (-k)
-log none/filename          : where to send log output (-l)
-mdi '<mdi flags>'          : pass flags to the MolSSI Driver Interface
-mpicolor color             : which exe in a multi-exe mpirun cmd (-m)
-cite                       : select citation reminder style (-c)
-nocite                     : disable citation reminder (-nc)
-nonbuf                     : disable screen/logfile buffering (-nb)
-package style ...          : invoke package command (-pk)
-partition size1 size2 ...  : assign partition sizes (-p)
-plog basename              : basename for partition logs (-pl)
-pscreen basename           : basename for partition screens (-ps)
-restart2data rfile dfile ... : convert restart to data file (-r2data)
-restart2dump rfile dgroup dstyle dfile ... 
                            : convert restart to dump file (-r2dump)
-restart2info rfile         : print info about restart rfile (-r2info)
-reorder topology-specs     : processor reordering (-r)
-screen none/filename       : where to send screen output (-sc)
-skiprun                    : skip loops in run and minimize (-sr)
-suffix gpu/intel/kk/opt/omp: style suffix to apply (-sf)
-var varname value          : set index style variable (-v)

OS: Linux "Ubuntu 20.04.2 LTS (fossa-ditto X84)" 5.15.0-124-generic x86_64

Compiler: GNU C++ 9.4.0 with OpenMP 4.5
C++ standard: C++11
MPI v3.1: Open MPI v4.0.3, package: Debian OpenMPI, ident: 4.0.3, repo rev: v4.0.3, Mar 03, 2020

Accelerator configuration:


FFT information:

FFT precision  = double
FFT engine  = mpiFFT
FFT library = FFTW3 with threads

Active compile time flags:

-DLAMMPS_GZIP
-DLAMMPS_JPEG
-DLAMMPS_FFMPEG
-DLAMMPS_SMALLBIG
sizeof(smallint): 32-bit
sizeof(imageint): 32-bit
sizeof(tagint):   32-bit
sizeof(bigint):   64-bit

Available compression formats:

Extension: .gz     Command: gzip
Extension: .bz2    Command: bzip2
Extension: .zst    Command: zstd
Extension: .xz     Command: xz
Extension: .lzma   Command: xz
Extension: .lz4    Command: lz4


Installed packages:

KSPACE MANYBODY MOLECULE REAXFF RIGID 
```

**Comment by mkanski** [comment]
Sorry, I misunderstood your comment, I thought the problem occurs on every step after some time. I can confirm the bug, below I'm attaching a simplified input which triggers it on 2 MPI threads. 

The issue appears when `fix reaxff/species` is triggered on the same step when an atom migrates from one MPI rank to another. I don't know enough about LAMMPS communication to fix it. One thing that goes around the bug is moving the species calculation from `post_integrate` to `end_of_step`, when all atoms are already migrated, but there may be some other nuances with this approach I'm not seeing.

```
processors * 1 1
units real
atom_style charge
boundary p p p

read_data data.test

pair_style reaxff NULL checkqeq no
pair_coeff * * cho.reax H C O

timestep 0.1
variable temp equal 1500
velocity all create ${temp} 10 dist gaussian

neigh_modify every 10 check no

fix species all reaxff/species 1 1 10 species.run element H C O
fix nvt all nvt temp ${temp} ${temp} $(500.0*dt)

dump dout all custom 10 dump.run id element q f_species x y z
dump_modify dout sort id element H C O

run 10

```

**Comment by navlalli** [comment]
I see, that's good to know. Thanks a lot for looking into this! Do you know if this can safely be moved to `end_of_step` @jrgissing?

**Comment by jrgissing** [comment]
@mkanski @navlalli thanks for identifying and troubleshooting this issue. yes, clusterIDs are currently not being communicated. i think moving to end_of_step would have been a good solution, except for the recent 'delete species' feature. So, I propose to properly communicate clusterID instead. Can you take a look at #4443 to see if it fixes this issue? Bonus points for testing with the 'delete' and 'delete_rate_limit' keywords as well

**Comment by navlalli** [comment]
@jrgissing thank you very much for getting this fixed.

The molecule IDs are now correct with #4443 for the example I provided above. However, there appears to be an issue with the molecule IDs when using the `delete` and `delete_rate_limit`

Using the exact same example as above with 
```
fix species all reaxff/species 1 1 1 species.run element ${atom1} ${atom2} ${atom3}
```
replaced by
```
fix species all reaxff/species 1 1 5 species.run element ${atom1} ${atom2} ${atom3} delete species.del masslimit 70 200
```
or 
```
fix species all reaxff/species 1 1 5 species.run element ${atom1} ${atom2} ${atom3} delete species.del masslimit 70 200 delete_rate_limit 2 20
```
The molecule IDs are all 0 for the first 4 timesteps (before any molecules are deleted). The molecules IDs seem more correct after this, but I'm not sure if the behaviour is exactly as intended. In some of the timesteps with deleted molecules, the atoms of the deleted molecules seem to be in the dump file. It may be the case that with the delete option, the molecules IDs are only getting updated every Nfreq timesteps?

Is the correct behaviour that non-deleted molecules should have molecule IDs from 1 to N, where N is the remaining number of molecules? It would be good to change the current documentation, which appears to be wrong.

![Image](https://github.com/user-attachments/assets/a9698a68-3b5a-4fb9-bfd7-43c1ad9e5911)

I've attached the example that reproduces the issue.

[test_delete.zip](https://github.com/user-attachments/files/18548032/test_delete.zip)

**Comment by jrgissing** [comment]
thanks for catching those issues. i pushed some new commits, do they result in expected behavior? my first instinct was to always index molecule IDs from 1 to Nmol, because I was using Nmol for easy testing, but I'm realizing it may be better to do what the original docs claimed, so that molecule IDs are persistent for use with MSD, etc. Please let me know what you think

**Comment by navlalli** [comment]
I tried out the new commits. It's great that this results in the correct molecule IDs on the first time step! The delete behaviour doesn't seem to be working as expected. For example, using 
```
fix species all reaxff/species 1 1 10 species.run element ${atom1} ${atom2} ${atom3} delete species.del masslimit 70 200 
```
leads to all 30 C12H26 molecules getting deleted straightaway. This is the line from the delete file:
```
Timestep 0 30 C12H26
```
Also when using 
```
fix species all reaxff/species 1 1 10 species.run element ${atom1} ${atom2} ${atom3} delete species.del masslimit 70 200 delete_rate_limit 2 20
```
the C12H26 molecules are deleted as expected. By timestep 300, all C12H26 molecules are deleted, but the molecule IDs of the oxygen molecules starts from 2, and not 1. After around 10 timesteps, the molecule IDs then correctly start from 1.

For the work I was doing, I simply needed a unique molecule ID for each molecule, so I was happy with the molecules being numbered from 1 to N. I liked this because it aligns with how you would normally label molecule IDs in non-reactive simulations. That being said, I don't have a particularly strong view on this. 

> so that molecule IDs are persistent for use with MSD, etc.

Perhaps I misunderstood your comment, but would this not be an issue in that a persistent molecule ID wouldn't necessarily correspond with the same molecule. For example, a hydrocarbon with molecule ID = 500 (the min atom ID of atoms in the hydrocarbon) could split into two fragments and one of those fragment would still have molecule ID = 500.

**Comment by jrgissing** [comment]
I agree, that molecule IDs should be reported on timestep 0, but deletions should probably not occur until timestep 1. Please let me know if the latest changes result in what you expected.

I think I was unable to reproduced the temporary incorrect molecule IDs at step 300, even before the latest change. How many processors are you using?

Regarding 1-Nmol vs max-atom-ID, I meant if you have a stable molecule, it would retain a consistent molID for its lifetime, for latter method.

**Comment by navlalli** [comment]
Latest changes seemed to have fixed both of those issues that I mentioned in the previous post. I think there may be a few other issues floating around. For example, the molecule ID 16 is missing at timestep 81 for around 9 timesteps. I've attached the files (see the dump file). I can't reproduce this exactly by running again, but I see similar issues every time. Essentially, I expect the max molecule ID at each timestep to be odd since I'm removing 2 molecules every 20 timesteps, but sometimes the last molecule ID is even. In these cases, there tends to be a molecule ID that is skipped. I'm guessing there is something stochastic somewhere that affects this, which may have been why you couldn't and I also can't reproduce the temporary incorrect molecule IDs at timestep 300. I am running in parallel with 16 cores, which you can see from run_sim.sh .

[missing-id.zip](https://github.com/user-attachments/files/18551768/missing-id.zip)

I reran a few times and can now see the molecule IDs starting at 2 at timestep 300. See the attached dump file.

[dump.zip](https://github.com/user-attachments/files/18551794/dump.zip)

> Regarding 1-Nmol vs max-atom-ID, I meant if you have a stable molecule, it would retain a consistent molID for its lifetime, for latter method.

Understood.
